### PR TITLE
Feature/fbfw 66 arrow

### PR DESCRIPTION
**Pull request by sariacodazen**
Needs touch ups for a couple of things but is a working product of drawing a scalable line between any two points given their id

Video Link of testing below

https://user-images.githubusercontent.com/95884853/156082353-d3e4f06c-3526-44ba-a0a9-46af57b6bbda.mov

Below is how I applied the drawArrow function. This is just a representation and most likely not how we will be using it as it will be utilized on fleets. This merely demonstrates a use case of connecting arrows to any two centers on the map
Screen Shot 2022-02-28 at 4 24 08 PM

### CODE REVIEW

File: beta-src/src/components/map/WDMapController.tsx [hunk 1]
```diff
@@ -63,7 +63,9 @@ const WDMapController: React.FC<WDMapControllerProps> = function ({
         .on("zoom", zoom);
 
       fullMap
-        .on("wheel", (e) => e.preventDefault())
+        .on("wheel", (e) => {
+          e.preventDefault();
+        })
```
Comment (nebevets) [inline beta-src/src/components/map/WDMapController.tsx:68]: i think you can safely revert this change

File: beta-src/src/utils/draw-arrow.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
+export default function drawArrow(actionType, source, svgElement, target) {
```
Comment (nebevets) [inline beta-src/src/utils/draw-arrow.ts:1]: there is not typing going on here. we need to define an interface, otherwise typescript is irrelevant.

File: beta-src/src/utils/draw-arrow.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  const actionTypeColors = {
+    moveOrder: "#FFFFFF",
+    move: "#000000",
+    moveConvoy: "#2042B8",
+    moveFailed: "#BB0000",
+    moveSupport: "#F8F83D",
+    holdSupport: "#3FC621",
+    retreat: "#BD2894",
+  };
```
Comment (nebevets) [inline beta-src/src/utils/draw-arrow.ts:10]: these should be in the theme or outside of this file in case we want to reuse them

File: beta-src/src/utils/draw-arrow.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    .select(`#${source}`)
+    .node()
+    .getBoundingClientRect();
+  const targetNodeData = svgElement
+    .select(`#${target}`)
+    .node()
+    .getBoundingClientRect();
+
+  svgElement.select("#arrowline").remove();
+  svgElement.select("#arrowhead").remove();
```
Comment (nebevets) [inline beta-src/src/utils/draw-arrow.ts:21]: why are we removing these? i don't think i understand what is going on here.

**Review by nebevets — CHANGES_REQUESTED** [review]

**Review by bryancodazen — CHANGES_REQUESTED** [review]
You need to run `npm run build` and apply those files to your PR. 

This PR needs a testing link (sandbox on netlify like https://webdip.netlify.app) or a video. 

You also need to provide example usage of this, I have no idea from looking at your PR how this can be used. 

Your testing link or video should show an example with multiple arrows drawn which scale to conform with zooming and panning. 

The file names in `utils` are using camel case. Please update `draw-arrow` to `drawArrow` and move it to the `map` directory which is being used for map related utilities.

File: beta-src/src/utils/draw-arrow.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    retreat: "#BD2894",
+  };
+  const sourceNodeData = svgElement
+    .select(`#${source}`)
+    .node()
+    .getBoundingClientRect();
+  const targetNodeData = svgElement
+    .select(`#${target}`)
+    .node()
+    .getBoundingClientRect();
```
[the remark this answers is not in the document]
Comment (sariacodazen) [reply beta-src/src/utils/draw-arrow.ts:18]: Using BBox returned the wrong coordinates, using boundingClientRect returned the right ones so I went with that

File: beta-src/src/utils/draw-arrow.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    .append("line")
+    .attr("x1", sourceNodeData.x + sourceNodeData.width / 2)
+    .attr("y1", sourceNodeData.y + sourceNodeData.height / 2)
+    .attr("x2", targetNodeData.x + targetNodeData.width / 2)
+    .attr("y2", targetNodeData.y + targetNodeData.height / 2)
+    .attr("marker-end", "url(#arrowhead)")
+    .attr("stroke", actionTypeColors[actionType])
+    .attr("stroke-width", "2")
+    .attr("id", "arrowline");
+}
```
[the remark this answers is not in the document]
Comment (sariacodazen) [reply beta-src/src/utils/draw-arrow.ts:46]: I was actually under the impression that only one arrow would be drawn at a time as I thought one user can plan one move at a time

File: beta-src/src/utils/draw-arrow.ts (hunk 4 above)
Comment (sariacodazen) [reply beta-src/src/utils/draw-arrow.ts:21]: These being removed allows for the arrow to be redrawn according to the right scaling. I initially thought only one arrow can appear at a time so I did not account for multiple arrows

File: beta-src/src/utils/draw-arrow.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (bryancodazen) [reply beta-src/src/utils/draw-arrow.ts:46]: @sariacodazen Arrows will be drawn one at a time when submitting moves but arrow drawing needs to run concurrently with the map re-drawing doesn't it? So we would need to consistently re-draw arrows based on saved moves / previous moves. 

Note that manually re-drawing is only based on the assumption that simply appending a def to the SVG would not scale with it. When you originally appended the arrow to the SVG, what specific element did you append it to?

File: beta-src/src/utils/drawArrow.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+   */
+  arrowColor: ArrowColors,
+  /**
+   * source is the id of the element the arrow begins at
+   */
+  source: string,
+  /**
+   * svg parent element; must be able to target centers and defs
+   */
+  svgElement: any,
```
Comment (nebevets) [inline beta-src/src/utils/drawArrow.ts:15]: any could just be anything. i could pass <div>anything</div> and it would not give me a typescript error, but this code would fail. i think you want SVGSVGElement

File: beta-src/src/utils/drawArrow.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+export default function drawArrow(
+  /**
+   * color code passed from enum ArrowColors based on move type
+   */
+  arrowColor: ArrowColors,
+  /**
+   * source is the id of the element the arrow begins at
+   */
+  source: string,
```
Comment (nebevets) [inline beta-src/src/utils/drawArrow.ts:11]: startElementID

File: beta-src/src/utils/drawArrow.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  /**
+   * svg parent element; must be able to target centers and defs
+   */
+  svgElement: any,
+  /**
+   * target is the id of the element the arrow ends at
+   */
+  target: string,
```
Comment (nebevets) [inline beta-src/src/utils/drawArrow.ts:19]: targetElementID

File: beta-src/src/utils/drawArrow.ts [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+) {
+  const arrowIdNumber = Math.floor(Date.now() + Math.random());
```
Comment (nebevets) [inline beta-src/src/utils/drawArrow.ts:21]: let's make the id unique by concatenating the two ids being passed in...

```
const arrowID = `${sourceElementID}-${targetElementID}`
```

File: beta-src/src/utils/removeArrow.ts [hunk 11]
```diff
@@ -0,0 +1,14 @@
+export default function removeArrow(
+  /**
+   * id is the id of the arrowline. This ends with a -somenumber
+   */
+  id: string,
+  /**
+   * svgElement is the parent svg element. Must be able to target defs as well as svg#container
+   */
+  svgElement: any,
```
Comment (nebevets) [inline beta-src/src/utils/removeArrow.ts:9]: never use any. i think it should be SVGSVGElement

File: beta-src/src/utils/removeArrow.ts [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+) {
+  const arrowIdTag = id.split("-")[1];
```
Comment (nebevets) [inline beta-src/src/utils/removeArrow.ts:11]:
```
const [_, arrowId] = id.split("-");
```

File: beta-src/src/utils/removeArrow.ts [hunk 13]
```diff
@@ -0,0 +1,14 @@
+export default function removeArrow(
+  /**
+   * id is the id of the arrowline. This ends with a -somenumber
+   */
+  id: string,
+  /**
+   * svgElement is the parent svg element. Must be able to target defs as well as svg#container
+   */
+  svgElement: any,
+) {
```
Comment (nebevets) [inline beta-src/src/utils/removeArrow.ts:10]: ): void {

File: beta-src/src/utils/drawArrow.ts [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+  source: string,
+  /**
+   * svg parent element; must be able to target centers and defs
+   */
+  svgElement: any,
+  /**
+   * target is the id of the element the arrow ends at
+   */
+  target: string,
+) {
```
Comment (nebevets) [inline beta-src/src/utils/drawArrow.ts:20]: ): void {

**Review by nebevets — CHANGES_REQUESTED** [review]
getting closer

File: beta-src/tsconfig.json [hunk 15]
```diff
@@ -1,11 +1,7 @@
 {
```
Comment (bryancodazen) [inline beta-src/tsconfig.json:1]: These seem like unnecessary changes to the file, how did they come about? This should probably be reverted.

File: beta-src/src/utils/drawArrow.ts [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    .attr("markerWidth", 8)
+    .attr("markerHeight", 8)
+    .attr("refX", 0)
+    .attr("refY", 4)
+    .attr("orient", "auto")
+    .append("polygon")
+    .attr("points", "0 0, 8 4, 0 8")
+    .attr("fill", arrowColor);
+
+  d3MapSelector
```
Comment (bryancodazen) [inline beta-src/src/utils/drawArrow.ts:41]: Why not just pass in the SVG element and then use the d3 selector in here like they are used in WDMapController so that types are inferred?

File: beta-src/src/enums/MoveOrder.ts [hunk 17]
```diff
@@ -0,0 +1,11 @@
+enum MoveOrder {
+  MOVE_ORDER_SELECTED = "moveOrderSelected",
```
Comment (bryancodazen) [inline beta-src/src/enums/MoveOrder.ts:2]: `MOVE_ORDER_SELECTED` isn't a "move" order, you should change this enum to just be `Order` or `OrderType`

File: beta-src/src/webDiplomacyTheme.ts [hunk 18]
```diff
@@ -103,6 +113,16 @@ const countryPalette: CountryPaletteOptions = {
   },
 };
 
+const arrowPalette: ArrowPalette = {
+  moveOrderSelected: "#FFFFFF",
```
Comment (bryancodazen) [inline beta-src/src/webDiplomacyTheme.ts:117]: What arrow are you drawing on `moveOrderSelected`? I'm not sure why this is here.

File: beta-src/src/components/map/WDMapController.tsx [hunk 19]
```diff
@@ -39,8 +39,10 @@ const WDMapController: React.FC<WDMapControllerProps> = function ({
     if (svgElement.current) {
       const fullMap = d3.select(svgElement.current);
       const contained = fullMap.select("#container");
-      const containedRect = contained.node().getBBox();
-      const gameBoardAreaRect = fullMap.select("#outlines").node().getBBox();
+      const containedRect = (contained.node() as SVGSVGElement).getBBox();
+      const gameBoardAreaRect = (
+        fullMap.select("#outlines").node() as SVGSVGElement
+      ).getBBox();
```
Comment (bryancodazen) [inline beta-src/src/components/map/WDMapController.tsx:45]: Isn't this pointless unless you're using @types/d3?

**Review by bryancodazen — CHANGES_REQUESTED** [review]

File: beta-src/src/utils/drawArrow.ts [hunk 20]
```diff
@@ -0,0 +1,52 @@
+export default function drawArrow(
+  /**
+   * color code passed from enum ArrowColors based on move type
+   */
+  arrowColor: string,
```
Comment (bryancodazen) [inline beta-src/src/utils/drawArrow.ts:5]: This shouldn't be a prop, the correct color should be deduced based on the order type inside this function. Using the current theme and receiving a proper order type enum.

File: beta-src/src/utils/drawArrow.ts (hunk 16 above)
Comment (sariacodazen) [reply beta-src/src/utils/drawArrow.ts:41]: This would require me to import d3 into the function file which I feel to be pointless if the controller of the draw arrow functionality will more than likely need to have d3 and have a reference to the MapControllers already declared d3 selected svg; this way we can just pass whats already declared.

File: beta-src/src/webDiplomacyTheme.ts (hunk 18 above)
Comment (sariacodazen) [reply beta-src/src/webDiplomacyTheme.ts:117]: This also answers the question above; the task shows selecting to move order, is indeed a move with a white arrow
Screen Shot 2022-02-10 at 1 15 39 PM

File: beta-src/src/components/map/WDMapController.tsx (hunk 19 above)
Comment (sariacodazen) [reply beta-src/src/components/map/WDMapController.tsx:45]: I was going to remove this when cleaning up, but figured it would be nice to force the SVG type for that node as it needs to be an svg for BBox if I am not mistaken

File: beta-src/src/utils/drawArrow.ts (hunk 20 above)
Comment (sariacodazen) [reply beta-src/src/utils/drawArrow.ts:5]: I was under the assumption that this task was to take in two ids and a color and draw the line? Wouldnt the user of the arrow drawing function use the palette I created and pass that value to my function?

File: beta-src/src/webDiplomacyTheme.ts [hunk 21]
```diff
@@ -103,6 +113,16 @@ const countryPalette: CountryPaletteOptions = {
...
+  move: "#000000",
+  convoy: "#2042B8",
+  moveFailed: "#BB0000",
+  moveSupport: "#F8F83D",
+  holdSupport: "#3FC621",
+  retreat: "#BD2894",
+};
```
[the remark this answers is not in the document]
Comment (sariacodazen) [reply beta-src/src/webDiplomacyTheme.ts:124]: Screen Shot 2022-03-02 at 4 29 35 PM
This indeed worked when I tested it; here is it being added to palette

File: beta-src/src/webDiplomacyTheme.ts (hunk 18 above)
Comment (bryancodazen) [reply beta-src/src/webDiplomacyTheme.ts:117]: @sariacodazen gotcha, maybe the enum `Arrow` or `ArrowType` makes more sense since we also have the move failed arrow.

File: beta-src/src/webDiplomacyTheme.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (bryancodazen) [reply beta-src/src/webDiplomacyTheme.ts:124]: I didn't say added to the palette, I said added to the palette **interface**.

File: beta-src/src/enums/MoveOrder.ts (hunk 17 above)
Comment (bryancodazen) [reply beta-src/src/enums/MoveOrder.ts:2]: Or maybe `Arrow` / `ArrowType` is better since move failed isn't really an order or a move either.

**Review by bryancodazen — CHANGES_REQUESTED** [review]

File: beta-src/src/webDiplomacyTheme.ts [hunk 22]
```diff
@@ -41,9 +41,26 @@ declare module "@mui/material/styles" {
     Russia: CountryColor;
     Italy: CountryColor;
     Turkey: CountryColor;
+    moveOrderSelected: string;
+    move: string;
+    convoy: string;
+    moveFailed: string;
+    moveSupport: string;
+    holdSupport: string;
+    retreat: string;
```
Comment (bryancodazen) [inline beta-src/src/webDiplomacyTheme.ts:50]: Move these into an `arrow` property or something so that they're not placed directly into the palette... 

Example usage: `theme.palette.arrowColors.holdSupport`

**Review by bryancodazen — CHANGES_REQUESTED** [review]

**Review by bryancodazen — APPROVED** [review]

**Review by nebevets — CHANGES_REQUESTED** [review]
we can sync on how to make the customer color interfaces re-usable

File: beta-src/tsconfig.json [hunk 23]
```diff
@@ -1,11 +1,7 @@
...
   "compilerOptions": {
     "target": "es5",
-    "lib": [
-      "dom",
-      "dom.iterable",
-      "esnext"
-    ],
+    "lib": ["dom", "dom.iterable", "esnext"],
```
[the remark this answers is not in the document]
Comment (sariacodazen) [reply beta-src/tsconfig.json:4]: This file was not formatted before so it got formatted now

**Review by nebevets — APPROVED** [review]
LGTM
